src/backend/distributed/planner/recursive_planning.c

```diff
@@ -316,6 +302,211 @@ RecursivelyPlanCTEs(Query *query, RecursivePlanningContext *planningContext)
 }
 
 
+/*
+ * RecursivelyPlanSubqueryWalker recursively finds all the Query nodes and
+ * recursively plans if necessary.
+ */
+static bool
+RecursivelyPlanSubqueryWalker(Node *node, RecursivePlanningContext *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		DeferredErrorMessage *error = NULL;
+
+		context->level += 1;
+		error = RecursivelyPlanSubqueriesAndCTEs(query, context);
+		if (error != NULL)
+		{
+			RaiseDeferredError(error, ERROR);
+		}
+		context->level -= 1;
+
+		if (ShouldRecursivelyPlanSubquery(query, context))
+		{
+			RecursivelyPlanSubquery(query, context);
+		}
+
+		return false;
+	}
+
+	return expression_tree_walker(node, RecursivelyPlanSubqueryWalker, context);
+}
+
+
+/*
+ * ShouldRecursivelyPlanSubquery decides whether the input subquery should be recursively
+ * planned or not.
+ *
+ * For the details, see the cases in the function.
+ */
+static bool
+ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context)
+{
+	bool shouldRecursivelyPlan = false;
+
+	if (FindNodeCheckInRangeTableList(subquery->rtable, IsLocalTableRTE))
```

Comment:
This might recurse too aggressively, e.g.:

```
SELECT * FROM distributed_table WHERE x IN (
  SELECT y FROM local_table JOIN (SELECT y FROM other_local_table GROUP BY y) USING (y)
);
```

We'd recursively plan the inner query on the local table, but then the outer query would break.

Reply:
We should maybe set `shouldRecursivelyPlan` to false instead and handle queries on local tables later.

I think recursive planning of subqueries is a 2-step process:
- decide if the query needs to be planned recursively based on its own properties (e.g. LIMIT)
- decide if any subqueries of the query need to be planned recursively based on the properties of the query (e.g. union, left join, subqueries on local tables in a distributed table query)

Reply:
When we discussed about re-partition joins in the subqueries, we thought it makes more sense to subplan them immediately at the leaf query as [here](https://github.com/citusdata/citus/compare/sql_monster#diff-1fb270ce40ed6550c916670330d5c788R1611). 

I thought this is a similar approach to that. I can see that the above query would fail. But, on the other hand, I'd like to handle local queries at the leaf level so that we don't need to adjust other parts of the planner to be safe for local tables.  So I'd like to think about more before changing this point.

Reply:
> When we discussed about re-partition joins in the subqueries, we thought it makes more sense to subplan them immediately at the leaf query as here.

There's no way we could support a repartition join in a subquery in any other way than recursively planning it, so we might as well do that immediately, especially because it improves the non-co-located join code. Once we have a smarter join order planner, we would also defer this decision.

> we don't need to adjust other parts of the planner to be safe for local tables

That's unavoidable for any query that has a local table in the FROM and a subquery that contains a distributed table, so keeping local tables away from the planner is not an objective.

I'm also fine with punting on supporting subqueries with local tables for now, it's not really a requirement.